pkg/object/autocertmanager/autocertmanager.go

```diff
@@ -0,0 +1,417 @@
+func (acm *AutoCertManager) DefaultSpec() interface{} {
+	return &Spec{
+		DirectoryURL:    acme.LetsEncryptURL,
+		RenewBefore:     "720h",
+		EnableHTTP01:    true,
+		EnableTLSALPN01: true,
+		EnableDNS01:     true,
+	}
+}
+
+// Init initializes AutoCertManager.
+func (acm *AutoCertManager) Init(superSpec *supervisor.Spec) {
+	acm.superSpec = superSpec
+	acm.spec = superSpec.ObjectSpec().(*Spec)
+	acm.super = superSpec.Super()
+
+	// TODO: remove this check after converting AutoCertManager to a system controller
+	if globalACM.Load() != nil {
+		logger.Warnf("an AutoCertManager instance is already exist")
+	}
+
+	acm.reload()
+}
+
+// Inherit inherits previous generation of AutoCertManager.
+func (acm *AutoCertManager) Inherit(superSpec *supervisor.Spec, previousGeneration supervisor.Object) {
+	acm.superSpec = superSpec
+	acm.spec = superSpec.ObjectSpec().(*Spec)
+	acm.super = superSpec.Super()
+
+	acm.reload()
+	previousGeneration.(*AutoCertManager).Close()
+}
+
+func (acm *AutoCertManager) findDomain(name string, exactMatch bool) *Domain {
+	// compare one by one, if many domains are configured, we need to refactor
+	// this function to improve performance.
+	for i := range acm.domains {
+		domain := &acm.domains[i]
+		if name == domain.Name {
+			return domain
+		}
+		if exactMatch || !domain.isWildcard() {
+			continue
+		}
+		if strings.HasSuffix(name, domain.Name[1:]) {
+			return domain
+		}
+	}
+	return nil
+}
+
+func (acm *AutoCertManager) reload() {
+	acm.stopCtx, acm.cancel = context.WithCancel(context.Background())
+	acm.storage = newStorage(acm.super.Cluster())
+
+	acm.renewBefore, _ = time.ParseDuration(acm.spec.RenewBefore)
+
+	acm.domains = make([]Domain, len(acm.spec.Domains))
+	for i := range acm.spec.Domains {
+		spec := &acm.spec.Domains[i]
+
+		// convert to puny code to support Chinese or other unicode domain names
+		name, err := idna.Lookup.ToASCII(spec.Name)
+		if err != nil && spec.Name[0] == '*' {
+			name, err = idna.Lookup.ToASCII(spec.Name[1:])
+			name = "*" + name
+		}
+		if err != nil {
+			name = spec.Name
+			logger.Warnf("domain name contains invalid characters: %s", name)
+		}
+
+		d := &acm.domains[i]
+		d.DomainSpec = spec
+		d.nameInPunyCode = name
+
+		cert, err := acm.storage.getCert(d.nameInPunyCode)
+		if err != nil {
+			logger.Infof("failed to load certificate for domain %s: %v", spec.Name, err)
+			continue
+		}
+		d.certificate.Store(cert)
+	}
+
+	globalACM.Store(acm)
+	go acm.run()
+	go acm.watchCertificate()
+}
+
+// Status returns the status of AutoCertManager.
+func (acm *AutoCertManager) Status() *supervisor.Status {
+	status := &Status{}
+	for i := range acm.domains {
+		d := &acm.domains[i]
+		status.Domains = append(status.Domains, CertificateStatus{
+			Name:       d.Name,
+			ExpireTime: d.certExpireTime(),
+		})
+	}
+	return &supervisor.Status{ObjectStatus: status}
+}
+
+// Close closes AutoCertManager.
+func (acm *AutoCertManager) Close() {
```

Comment:
When close, do we need to delete the database?

Reply:
In `storage.go` we put some data into cluster (etcd). When do `inherit`, it's ok. But when customer delete the object, `Close` method will be called, do we need to delete all data that put into etcd?

Reply:
`Close` is also called at program exit. And some CAs may have a limit of certificate renewal rate, so better to keep the data as it may be reused later.